packages/test-base/src/androidTest/kotlin/io/realm/kotlin/test/shared/RealmInMemoryTests.kt

```diff
@@ -0,0 +1,243 @@
+package io.realm.kotlin.test.shared
+
+import io.realm.kotlin.Configuration
+import io.realm.kotlin.Realm
+import io.realm.kotlin.RealmConfiguration
+import io.realm.kotlin.entities.Sample
+import io.realm.kotlin.ext.query
+import io.realm.kotlin.internal.RealmImpl
+import io.realm.kotlin.internal.platform.runBlocking
+import io.realm.kotlin.test.platform.PlatformUtils
+import io.realm.kotlin.test.util.TestHelper
+import junit.framework.AssertionFailedError
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.TimeUnit
+import kotlin.random.Random
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+import kotlin.test.fail
+
+class RealmInMemoryTests {
+    private lateinit var tmpDir: String
+    private lateinit var realm: Realm
+    private lateinit var inMemConf: Configuration
+    private lateinit var onDiskConf: Configuration
+    @Before
+    fun setup() {
+        tmpDir = PlatformUtils.createTempDir()
+        inMemConf = RealmConfiguration.Builder(schema = setOf(Sample::class))
+            .directory(tmpDir)
+            .inMemory()
+            .build()
+        onDiskConf = RealmConfiguration.Builder(schema = setOf(Sample::class))
+            .directory(tmpDir)
+            .build()
+        realm = Realm.open(inMemConf)
+    }
+
+    @After
+    fun tearDown() {
+        PlatformUtils.deleteTempDir(tmpDir)
+        if (realm != null) {
+            realm.close()
+        }
+    }
+
+    @Test
+    fun inMemoryRealm_wipedOnClose() {
+        realm.writeBlocking {
+            copyToRealm(Sample())
+        }
+        assertEquals(1, realm.query(Sample::class).count().find())
+        realm.close()
+        realm = Realm.open(inMemConf)
+        assertEquals(0, realm.query(Sample::class).count().find())
+    }
+
+    @Test
+    fun inMemoryRealm_withDifferentNames() {
+        realm.writeBlocking {
+            copyToRealm(Sample().apply { stringField = "foo" })
+        }
+
+        // Creates the 2nd in-memory Realm with a different name. To make sure they are not affecting each other.
+        val tmpDir2 = PlatformUtils.createTempDir("2")
+        val configuration2 = RealmConfiguration.Builder(schema = setOf(Sample::class))
+            .directory(tmpDir2)
+            .inMemory()
+            .build()
+        val realm2 = Realm.open(configuration2)
+        realm2.writeBlocking {
+            copyToRealm(Sample().apply { stringField = "bar" })
+        }
+        assertEquals(1, realm.query(Sample::class).count().find())
+        assertEquals("foo", realm.query<Sample>("stringField == 'foo'").find().first().stringField)
+        assertEquals(1, realm2.query(Sample::class).count().find())
+        assertEquals("bar", realm2.query<Sample>("stringField == 'bar'").find().first().stringField)
+        realm2.close()
+        PlatformUtils.deleteTempDir(tmpDir2)
```

Comment:
I would move this into a `try { } finally { }` block with the asserts inside try and cleanup inside `finally`. That way we will always do cleanup even if the test fails.